aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,344 @@
+				mgr.logger.WithFields(
+					xlog.NewLogField("shard", shard),
+				).Warn("shard not found in flush times")
+				continue
+			}
+			resolution := flusher.Resolution()
+			lastFlushedAtNanos, exists := shardFlushTimes.ByResolution[int64(resolution)]
+			if !exists {
+				mgr.metrics.resolutionNotFound.Inc(1)
+				mgr.logger.WithFields(
+					xlog.NewLogField("shard", shard),
+					xlog.NewLogField("resolution", resolution.String()),
+				).Warn("resolution not found in flush times")
+				continue
+			}
+			newFlushTarget := flusherWithTime{
+				flusher:          flusher,
+				flushBeforeNanos: lastFlushedAtNanos,
+			}
+			flushersByInterval[i].flushers = append(flushersByInterval[i].flushers, newFlushTarget)
+		}
+	}
+	return flushersByInterval
+}
+
+func (mgr *followerFlushManager) flushersFromForcedFlushWithLock(
+	buckets []*flushBucket,
+	flushBeforeNanos int64,
+) []flushersGroup {
+	flushersByInterval := make([]flushersGroup, len(buckets))
+	for i, bucket := range buckets {
+		flushersByInterval[i].interval = bucket.interval
+		flushersByInterval[i].duration = bucket.duration
+		flushersByInterval[i].flushers = make([]flusherWithTime, 0, defaultInitialFlushTimesCapacity)
+		for _, flusher := range bucket.flushers {
+			newFlushTarget := flusherWithTime{
+				flusher:          flusher,
+				flushBeforeNanos: flushBeforeNanos,
+			}
+			flushersByInterval[i].flushers = append(flushersByInterval[i].flushers, newFlushTarget)
+		}
+	}
+	return flushersByInterval
+}
+
+func (mgr *followerFlushManager) watchFlushTimes() {
+	var (
+		throttlePeriod  = time.Second
+		flushTimesWatch kv.ValueWatch
+		err             error
+	)
+
+	for {
+		if flushTimesWatch == nil {
+			flushTimesWatch, err = mgr.flushTimesStore.Watch(mgr.flushTimesKey)
+			if err != nil {
+				mgr.metrics.watchCreateErrors.Inc(1)
+				mgr.logger.WithFields(
+					xlog.NewLogField("flushTimesKey", mgr.flushTimesKey),
+					xlog.NewLogErrField(err),
+				).Error("error creating flush times watch")
+				mgr.sleepFn(throttlePeriod)
+				continue
+			}
+		}
+
+		select {
+		case <-flushTimesWatch.C():
+		case <-mgr.doneCh:
+			return
+		}
+
+		value := flushTimesWatch.Get()
+		mgr.Lock()
+		mgr.proto.Reset()
+		if err = value.Unmarshal(&mgr.proto); err != nil {
+			mgr.metrics.unmarshalErrors.Inc(1)
+			mgr.logger.WithFields(
+				xlog.NewLogField("flushTimesKey", mgr.flushTimesKey),
+				xlog.NewLogErrField(err),
+			).Error("flush times unmarshal error")
+			mgr.Unlock()
+			continue
+		}
+		mgr.flushTimesState = flushTimesUpdated
+		mgr.Unlock()
+	}
+}
+
+type followerFlushTask struct {
+	mgr                *followerFlushManager
+	flushersByInterval []flushersGroup
+}
+
+func (t *followerFlushTask) Run() {
+	var (
+		mgr       = t.mgr
+		wgWorkers sync.WaitGroup
+	)
+	for _, group := range t.flushersByInterval {
+		start := mgr.nowFn()
+		for _, flusherWithTime := range group.flushers {
+			flusherWithTime := flusherWithTime
+			wgWorkers.Add(1)
+			mgr.workers.Go(func() {
```

Comment:
That won't happen because the flushes are performed sequentially, i.e., the next flush won't even start until the current flush finishes, so the goroutines will always be returned back to pool before starting the next flush.